cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -0,0 +1,274 @@
+            proto: The `DeviceSpecification` proto describing a Google device.
+
+        Raises:
+            ValueError: If the given `DeviceSpecification` is invalid.
+        """
+
+        _validate_device_specification(proto)
+
+        # Create qubit set
+        all_qubits = [_qid_from_str(q) for q in proto.valid_qubits]
+
+        # Create qubit pair set
+        #
+        # While the `GateSpecification` proto message contains qubit target references, they are
+        # ignored here because the following assumptions make them unnecessary currently:
+        # * All valid qubit pairs work for all two-qubit gates.
+        # * All valid qubits work for all single-qubit gates.
+        # * Measurement gate can always be applied to all subset of qubits.
+        #
+        qubit_pairs = [
+            (_qid_from_str(target.ids[0]), _qid_from_str(target.ids[1]))
+            for ts in proto.valid_targets
+            for target in ts.targets
+            if len(target.ids) == 2 and ts.target_ordering == v2.device_pb2.TargetSet.SYMMETRIC
+        ]
+
+        # TODO(#5050) implement gate durations
+        try:
+            metadata = cirq.GridDeviceMetadata(
+                qubit_pairs=qubit_pairs,
+                gateset=cirq.Gateset(),  # TODO(#5050) implement
+                all_qubits=all_qubits,
+            )
+        except ValueError as ve:  # coverage: ignore
+            # Spec errors should have been caught in validation above.
+            raise ValueError("DeviceSpecification is invalid.") from ve  # coverage: ignore
+
+        return GridDevice(metadata)
+
+    @property
+    def metadata(self) -> cirq.GridDeviceMetadata:
+        """Get metadata information for the device."""
+        return self._metadata
+
+    def validate_operation(self, operation: cirq.Operation) -> None:
+        """Raises an exception if an operation is not valid.
+
+        An operation is valid if
+            * The operation is in the device gateset.
+            * The operation targets a valid qubit
+            * The operation targets a valid qubit pair, if it is a two-qubit operation.
+
+        Args:
+            operation: The operation to validate.
+
+        Raises:
+            ValueError: The operation isn't valid for this device.
+        """
+        # TODO(#5050) uncomment once gateset logic is implemented
+        # if operation not in self._metadata.gateset:
+        #     raise ValueError(f'Operation {operation} is not a supported gate')
+
+        for q in operation.qubits:
+            if q not in self._metadata.qubit_set:
+                raise ValueError(f'Qubit not on device: {q!r}')
+
+        if (
+            len(operation.qubits) == 2
+            and frozenset(operation.qubits) not in self._metadata.qubit_pairs
+        ):
+            raise ValueError(f'Qubit pair is not valid on device: {operation.qubits!r}')
+
+    def __str__(self) -> str:
+        # If all qubits are grid qubits, render an appropriate text diagram.
```

Comment:
I am surprised that the below code for rendering a grid device doesn't exist somewhere already.  Is the default insufficient because we do not assume pairwise connectivity between qubits?

Reply:
I got this code from [`SerializableDevice`](https://github.com/quantumlib/Cirq/blob/4ff626477bd93a81285c5673cc56044822eaa8af/cirq-google/cirq_google/devices/serializable_device.py#L214), and there's also something similar for [`XmonDevice`](https://github.com/quantumlib/Cirq/blob/4ff626477bd93a81285c5673cc56044822eaa8af/cirq-google/cirq_google/devices/xmon_device.py#L223). This `GridDevice` will eventually replace both of them so I don't think there's a need to define a common function.

`cirq.Device` doesn't have a default, and I presume it's because yeah connectivity can be more general and would be hard to render. cc @MichaelBroughton